Controller/Payment/Notification.php

```diff
@@ -3,130 +3,286 @@
-             */
-            if (!$isFullRefund && $order->getStatus() === Order::STATE_PROCESSING) {
-                $partialRefundNote = $note_prefix . 'Partial Refunded: ' . $refund_amount . '  |  Refund-Key: ' . $refund_key . '  |  Reason: ' . $refund_reason;
-                $order->addStatusToHistory(Order::STATE_PROCESSING, $partialRefundNote, false);
+                    /** Check order-id is not contain multishipping */
+                    if (strpos($midtransOrderId, 'multishipping-') !== true) {
+                        $order = $this->getOrderByIncrementId($midtransOrderId);
+                        $this->processRefund($order, $refunds, false, $grossAmount);
+                    }
+                }
             }
         }
+
         $this->saveOrder($order);
 
         /**
          * If log request isEnabled, add request payload to var/log/midtrans/request.log
          */
-        $_info = "status : " . $transaction . " , order : " . $orderId . ", payment type : " . $payment_type;
+        $_info = "status : " . $transaction . " , order : " . $midtransOrderId . ", payment type : " . $payment_type;
         $this->_midtransLogger->midtransNotification($_info);
     }
+
+    /**
+     * Handling refund from Magento dashboard and Midtrans MAP
+     *
+     * @param Order $orderRefund
+     * @param $refunds
+     * @param $isFromMagento
+     * @param null $grossAmount
+     * @throws \Exception
+     */
+    private function processRefund(Order $orderRefund, $refunds, $isFromMagento, $grossAmount = null)
+    {
+        $refund[] = end($refunds);
+        $refundAmount = $refund[0]->refund_amount;
+        $refund_reason = $refund[0]->reason;
+
+        $isFullRefund = $this->isFullRefund($refunds, $orderRefund, $isFromMagento, $grossAmount);
+        $refund_note = 'MIDTRANS NOTIFICATION | Refunded: ' . $refundAmount . '  |  Reason: ' . $refund_reason;
+
+        /** Handling full refund */
+        if ($isFullRefund && $orderRefund->getStatus() != Order::STATE_CLOSED && $orderRefund->getState() != Order::STATE_CLOSED) {
+            $this->cancelOrder($orderRefund, Order::STATE_CLOSED, $refund_note);
+        }
+        /** Handling partial refund */
+        elseif ($orderRefund->getStatus() != Order::STATE_CLOSED && $orderRefund->getState() != Order::STATE_CLOSED) {
+            /** Do not process if notif history already exist */
+            if (!$this->isOrderCommentExist($orderRefund, $refund_note)) {
+                if ($isFullRefund) {
+                    /** Close order if total amount refund array is equal with grand total order / gross amount */
+                    $this->cancelOrder($orderRefund, Order::STATE_CLOSED, $refund_note);
+                } else {
+                    /** Put status history if total amount refund array is not equal with grand total order / gross amount */
+                    $this->setOrderStateAndStatus($orderRefund, Order::STATE_PROCESSING, $refund_note);
```

Comment:
Not for now (maybe for future release): 
something to think about, is `STATE_PROCESSING` correct for refunded state? 
Sample case customer buy a Shoe, paid (state `processing`), then customer get the Shoe (state `shipped`), then customer ask for partial refund (state is back to `processing` modified by this Code). `processing` state maybe confusing for merchant, because merchant sure that the order has been `shipped`, why it got back into `processing` again?

Reply:
Noted, will add to the backlog